src/NuGet.Core/NuGet.Packaging/Rules/DependenciesGroupsForEachTFMRule.cs

```diff
@@ -0,0 +1,174 @@
+            return issues;
+        }
+
+        internal (HashSet<NuGetFramework>, HashSet<NuGetFramework>, HashSet<NuGetFramework>) Categorize(IEnumerable<string> files, IEnumerable<NuGetFramework> tfmsFromNuspec)
+        {
+            var managedCodeConventions = new ManagedCodeConventions(new RuntimeGraph());
+            Func<object, object, bool> isCompatible = managedCodeConventions.Properties["tfm"].CompatibilityTest;
+            var collection = new ContentItemCollection();
+            collection.Load(files);
+
+            var libItems = ContentExtractor.GetContentForPattern(collection, managedCodeConventions.Patterns.CompileLibAssemblies);
+            var refItems = ContentExtractor.GetContentForPattern(collection, managedCodeConventions.Patterns.CompileRefAssemblies);
+
+            var tfmsFromFilesSet = new HashSet<NuGetFramework>();
+            tfmsFromFilesSet.AddRange(ContentExtractor.GetGroupFrameworks(libItems));
+            tfmsFromFilesSet.AddRange(ContentExtractor.GetGroupFrameworks(refItems));
+
+            var tfmsFromFiles = tfmsFromFilesSet.Where(t => t != Net00).ToList();
+
+            var noExactMatchesFromFile = new HashSet<NuGetFramework>();
+            var noExactMatchesFromNuspec = new HashSet<NuGetFramework>();
+            var compatNotExactMatches = new HashSet<NuGetFramework>();
+
+            noExactMatchesFromFile.AddRange(tfmsFromFiles.Where(e => !tfmsFromNuspec.Contains(e)));
+            noExactMatchesFromNuspec.AddRange(tfmsFromNuspec.Where(e => !tfmsFromFiles.Contains(e)));
+            foreach (var fileTFM in noExactMatchesFromFile)
+            {
+                foreach (var nuspecTFM in noExactMatchesFromNuspec)
+                {
+                    if (isCompatible(fileTFM, nuspecTFM))
+                    {
+                        compatNotExactMatches.Add(fileTFM);
+                        break;
+                    }
+                }
+            }
+            noExactMatchesFromFile.RemoveWhere(m => compatNotExactMatches.Contains(m));
+            
+            return (compatNotExactMatches, noExactMatchesFromFile, noExactMatchesFromNuspec);
+        }
+        
+        internal (string noExactMatchString, string compatMatchString) GenerateWarningString(ICollection<NuGetFramework> noExactMatchesFromFile,
+                ICollection<NuGetFramework> noExactMatchesFromNuspec, ICollection<NuGetFramework> compatNotExactMatches)
+        {
+            var noExactMatchString = new StringBuilder();
+            var compatMatchString = new StringBuilder();
+
+            if (noExactMatchesFromFile.Count != 0)
+            {
+                foreach (var tfm in noExactMatchesFromFile)
+                {
+                    if(tfm == noExactMatchesFromFile.First())
+                    { 
+                        noExactMatchString.Append(AnalysisResources.DependenciesGroupsForEachTFMBeginningToNuspec);
+                        noExactMatchString.Append(" " + tfm.GetFrameworkString());
```

Comment:
Any time you use `+` on a string, it creates a temporary string that needs to be garbage collected. Use `.Append(" ")` and `.Append(tfm.GetFrameworkString())` as separate lines. Same goes throughout the file.